crypto/rsa/rsa_gen.c

```diff
@@ -66,9 +67,19 @@ static int rsa_builtin_keygen(RSA *rsa, int bits, int primes, BIGNUM *e_value,
                               BN_GENCB *cb)
 {
 #ifdef FIPS_MODE
+    int ret;
+    OSSL_CALLBACK *cb = NULL;
+    void *cbarg = NULL;
+
     if (primes != 2)
         return 0;
-    return rsa_sp800_56b_generate_key(rsa, bits, e_value, cb);
+    ret = rsa_sp800_56b_generate_key(rsa, bits, e_value, cb);
+    if (ret) {
+        /* TODO(3.0) - needs libctx */
+        OSSL_SELF_TEST_get_callback(NULL, &cb, &cbarg);
+        ret = SELF_TEST_keygen_pairwise_test_rsa(rsa, cb, cbarg);
+    }
```

Comment:
This function already takes a BN_GENCB, why not use it?

Reply:
see below